src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp

```diff
@@ -0,0 +1,145 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/IndexIterator.hpp"  // IWYU pragma: keep
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/ModalVector.hpp"
+#include "Domain/Mesh.hpp"  // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace {
+// Given the modal expansion corresponding to some function f(x), compute the
+// modal expansion corresponding to the derivative d/dx f(x).
+ModalVector modal_derivative(const ModalVector& coeffs) noexcept {
+  const size_t number_of_modes = coeffs.size();
+  ModalVector deriv_coeffs(number_of_modes, 0.);
+  for (size_t i = 0; i < number_of_modes; ++i) {
+    for (size_t j = (i % 2 == 0 ? 1 : 0); j < i; j += 2) {
+      deriv_coeffs[j] += coeffs[i] * (2. * j + 1.);
```

Comment:
not for this PR, but perhaps ModalVector should be templated on its basis?

Reply:
Yea, though that would be moving towards `DataMesh` :man_shrugging:

Reply:
If so, we should also template `DataVector`, because adding data from different grid points is just as meaningless as adding modes in different bases. (Am I just rephrasing what you are saying, @nilsdeppe?)

Reply:
Well I was asking @kidder what he thinks because I'm not sure what I think yet :P

Reply:
Never mind, I think the same reasons we decided not to template Mesh on the Basis will apply.
Perhaps the function should take the Basis and switch on it, with an ERROR for unimplemented